python/lsst/pipe/base/tests.py

```diff
@@ -0,0 +1,192 @@
+    butler : `lsst.daf.butler.Butler`
+        The collection the quantum refers to.
+    dataIds : `collections.abc.Mapping` [`str`]
+        A mapping keyed by input/output names. Values must be data IDs for
+        single connections and sequences of data IDs for multiple connections.
+
+    Returns
+    -------
+    quantum : `lsst.daf.butler.Quantum`
+        A quantum for ``task``, when called with ``dataIds``.
+    """
+    quantum = Quantum(taskClass=type(task))
+    connections = task.config.ConnectionsClass(config=task.config)
+
+    try:
+        for name in connections.inputs:
+            connection = connections.__getattribute__(name)
+            ids = _checkDataIdMultiplicity(dataIds[name], connection.multiple)
+            for id in ids:
+                quantum.addPredictedInput(_refFromConnection(butler, connection, id))
+        for name in connections.outputs:
+            connection = connections.__getattribute__(name)
+            ids = _checkDataIdMultiplicity(dataIds[name], connection.multiple)
+            for id in ids:
+                quantum.addOutput(_refFromConnection(butler, connection, id))
+        return quantum
+    except KeyError as e:
+        raise ValueError("Mismatch in input data.") from e
+
+
+def _checkDataIdMultiplicity(dataIds, multiple):
+    if multiple:
+        if isinstance(dataIds, collections.abc.Sequence):
+            return dataIds
+        else:
+            raise ValueError(f"Expected multiple data IDs, got {dataIds}.")
+    else:
+        if isinstance(dataIds, collections.abc.Mapping) or isinstance(dataIds, DataCoordinate):
+            return [dataIds]
+        else:
+            raise ValueError(f"Expected single data ID, got {dataIds}.")
+
+
+def _refFromConnection(butler, connection, dataId, **kwargs):
+    """Create a DatasetRef for a connection in a collection.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The collection to point to.
+    connection : `lsst.pipe.base.connectionTypes.DimensionedConnection`
+        The connection defining the dataset type to point to.
+    dataId
+        The data ID for the dataset to point to.
+    **kwargs
+        Additional keyword arguments used to augment or construct
+        a `~lsst.daf.butler.DataCoordinate`.
+
+    Returns
+    -------
+    ref : `lsst.daf.butler.DatasetRef`
+        A reference to a dataset compatible with ``connection``, with ID
+        ``dataId``, in the collection pointed to by ``butler``.
+    """
+    universe = butler.registry.dimensions
+    dataId = DataCoordinate.standardize(dataId, **kwargs, universe=universe)
+    datasetType = connection.makeDatasetType(universe)
+    try:
+        butler.registry.getDatasetType(datasetType.name)
+    except KeyError:
+        raise ValueError(f"Invalid dataset type {connection.name}.")
+    try:
+        return DatasetRef(
+            datasetType=datasetType,
+            dataId=dataId,
+        )
```

Comment:
I think it would be a bit clearer to explicitly define `ref = DatasetRef(stuff)` and then return `ref`.